File: src/main/store/accounts.ts
import { safeStorage } from 'electron'
import type { PublicAccount } from '@shared/types'
import { accountsFile } from '../paths'
import { readJsonFile, writeJsonFile } from './json'

export interface TokenBundle {
  msAccessToken: string
  msRefreshToken: string
  msalCache?: string
  msExpiresAt: number
  mcAccessToken: string
  mcExpiresAt: number
}

export interface StoredAccount extends PublicAccount {
  tokenBlob: string
}

interface AccountStore {
  activeAccountId: string | null
  accounts: StoredAccount[]
}

let cache: AccountStore | null = null

function emptyStore(): AccountStore {
  return { activeAccountId: null, accounts: [] }
}

export function encryptSecret(value: string): string {
  if (safeStorage.isEncryptionAvailable()) {
    return `enc:${safeStorage.encryptString(value).toString('base64')}`
  }
  return `plain:${Buffer.from(value, 'utf8').toString('base64')}`
}

export function decryptSecret(value: string): string {
  if (value.startsWith('enc:')) {
    return safeStorage.decryptString(Buffer.from(value.slice(4), 'base64'))
  }
  if (value.startsWith('plain:')) {
    return Buffer.from(value.slice(6), 'base64').toString('utf8')
  }
  return value
}

async function load(): Promise<AccountStore> {
  if (cache) return cache
  cache = await readJsonFile<AccountStore>(accountsFile(), emptyStore())
  return cache
}

async function persist(): Promise<void> {
  await writeJsonFile(accountsFile(), await load())
}

function toPublic(account: StoredAccount): PublicAccount {
  return {
    id: account.id,
    username: account.username,
    xboxGamertag: account.xboxGamertag,
    avatarUrl: account.avatarUrl
  }
}

export async function listPublicAccounts(): Promise<{
  accounts: PublicAccount[]
  activeAccountId: string | null
}> {
  const data = await load()
  return {
    activeAccountId: data.activeAccountId,
    accounts: data.accounts.map(toPublic)
  }
}

export async function getStoredAccount(id: string): Promise<StoredAccount> {
  const found = (await load()).accounts.find((item) => item.id === id)
  if (!found) throw new Error('Account not found.')
  return found
}

export async function getActiveAccount(): Promise<StoredAccount | null> {
  const data = await load()
  if (!data.activeAccountId) return data.accounts[0] ?? null
  return data.accounts.find((item) => item.id === data.activeAccountId) ?? data.accounts[0] ?? null
}

export async function upsertAccount(
  account: PublicAccount,
  tokens: TokenBundle
): Promise<PublicAccount> {
  const data = await load()
  const stored: StoredAccount = {
    ...account,
    tokenBlob: encryptSecret(JSON.stringify(tokens))
  }
  const index = data.accounts.findIndex((item) => item.id === account.id)
  if (index >= 0) data.accounts[index] = stored
  else data.accounts.unshift(stored)
  data.activeAccountId = account.id
  await persist()
  return toPublic(stored)
}

export async function readTokens(id: string): Promise<TokenBundle> {
  const stored = await getStoredAccount(id)
  return JSON.parse(decryptSecret(stored.tokenBlob)) as TokenBundle
}

export async function selectAccount(id: string): Promise<void> {
  const data = await load()
  if (!data.accounts.some((item) => item.id === id)) throw new Error('Account not found.')
  data.activeAccountId = id
  await persist()
}

export async function removeAccount(id: string): Promise<void> {
  const data = await load()
  data.accounts = data.accounts.filter((item) => item.id !== id)
  if (data.activeAccountId === id) {
    data.activeAccountId = data.accounts[0]?.id ?? null
  }
  await persist()
}

File: src-tauri/src/accounts.rs
use crate::crypto::{decrypt_secret, encrypt_secret};
use crate::error::{Result, WoolyError};
use crate::json::{read_json_file, write_json_file};
use crate::model::{AccountList, PublicAccount};
use crate::paths::accounts_file;
use serde::{Deserialize, Serialize};
use std::sync::Mutex;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TokenBundle {
    pub ms_access_token: String,
    pub ms_refresh_token: String,
    #[serde(default)]
    pub msal_cache: Option<String>,
    pub ms_expires_at: u64,
    pub mc_access_token: String,
    pub mc_expires_at: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct StoredAccount {
    id: String,
    username: String,
    #[serde(default)]
    xbox_gamertag: Option<String>,
    avatar_url: String,
    token_blob: String,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct AccountStore {
    active_account_id: Option<String>,
    #[serde(default)]
    accounts: Vec<StoredAccount>,
}

static CACHE: Mutex<Option<AccountStore>> = Mutex::new(None);

fn to_public(account: &StoredAccount) -> PublicAccount {
    PublicAccount {
        id: account.id.clone(),
        username: account.username.clone(),
        xbox_gamertag: account.xbox_gamertag.clone(),
        avatar_url: account.avatar_url.clone(),
    }
}

async fn load() -> Result<AccountStore> {
    if let Some(cached) = CACHE.lock().ok().and_then(|g| g.clone()) {
        return Ok(cached);
    }
    let data = read_json_file(&accounts_file(), AccountStore::default()).await;
    if let Ok(mut guard) = CACHE.lock() {
        *guard = Some(data.clone());
    }
    Ok(data)
}

async fn persist(data: &AccountStore) -> Result<()> {
    write_json_file(&accounts_file(), data).await?;
    if let Ok(mut guard) = CACHE.lock() {
        *guard = Some(data.clone());
    }
    Ok(())
}

pub async fn list_public_accounts() -> Result<AccountList> {
    let data = load().await?;
    Ok(AccountList {
        active_account_id: data.active_account_id,
        accounts: data.accounts.iter().map(to_public).collect(),
    })
}

pub async fn get_active_account() -> Result<Option<PublicAccount>> {
    let data = load().await?;
    let id = data.active_account_id.clone();
    let found = if let Some(id) = id {
        data.accounts
            .iter()
            .find(|a| a.id == id)
            .or(data.accounts.first())
    } else {
        data.accounts.first()
    };
    Ok(found.map(to_public))
}

pub async fn upsert_account(account: PublicAccount, tokens: TokenBundle) -> Result<PublicAccount> {
    let mut data = load().await?;
    let stored = StoredAccount {
        id: account.id.clone(),
        username: account.username.clone(),
        xbox_gamertag: account.xbox_gamertag.clone(),
        avatar_url: account.avatar_url.clone(),
        token_blob: encrypt_secret(&serde_json::to_string(&tokens)?),
    };
    if let Some(existing) = data.accounts.iter_mut().find(|a| a.id == account.id) {
        *existing = stored;
    } else {
        data.accounts.insert(0, stored);
    }
    data.active_account_id = Some(account.id.clone());
    persist(&data).await?;
    Ok(account)
}

pub async fn read_tokens(id: &str) -> Result<TokenBundle> {
    let data = load().await?;
    let stored = data
        .accounts
        .iter()
        .find(|a| a.id == id)
        .ok_or_else(|| WoolyError::msg("Account not found."))?;
    let json = decrypt_secret(&stored.token_blob)?;
    Ok(serde_json::from_str(&json)?)
}

pub async fn select_account(id: &str) -> Result<()> {
    let mut data = load().await?;
    if !data.accounts.iter().any(|a| a.id == id) {
        return Err(WoolyError::msg("Account not found."));
    }
    data.active_account_id = Some(id.to_string());
    persist(&data).await
}

pub async fn remove_account(id: &str) -> Result<()> {
    let mut data = load().await?;
    data.accounts.retain(|a| a.id != id);
    if data.active_account_id.as_deref() == Some(id) {
        data.active_account_id = data.accounts.first().map(|a| a.id.clone());
    }
    persist(&data).await
}
